### ISSUE TITLE
Terraforged / Ice & Fire Crash issue, possible fixes (1.16.5)

### ISSUE THREAD

**Issue by Kowantify** [Teneted/Tenet#969]
#816  #735 

I am having this EXACT same issue, to fix the terraforged issue, terraforged has a config file that has a Max Tick Time for loading chunks/structures. Increasing that by a large amount fixes the generating new world issue. 

Ice & Fire still crashes when flying around thanks to it reaching minecrafts server.properties built in tick time, which is affected by mohist in some way, as in its overwriting there max tick time threshold. although from what I remember there is a max tick time setting. Settings the server.properties max tick time to -1 MIGHT fix this issue entirely, but we shouldn't have to do that to begin with, this issue only happens when mohist is installed, I tested this on multiple versions. as long as mohist isn't the main .jar running the server it won't have these issues present at all, I also did not have ANY plugins installed, it was bare bones Mohist with those 2 mods

**Comment by joaoespanhol** [comment]
this is not a solution, because it will even end up with server tps, and when you have multiple players playing, it will crasher.

**Comment by Kowantify** [comment]
That is not how TPS works... Yes it will cause some lag, but that lag is going to be present with the tick limit and without, trust me i've tested this multiple times. Your going to get the same TPS loss with and without it, the issue is the server is killing itself because its "tick limit" for allowing a structure to generate was exceeded. If your lagging and losing TPS when chunks are generated and it gets worse with multiple players, it's a server performance issue, as in hardware can't keep up issue. Not a "Tick" issue

**Comment by Dokman** [comment]
i am having the same problem but with another 3 mods in my server. i have no plugins too.

**Comment by Kowantify** [comment]
I think this issue isn't just Terraforged and Ice & Fire, I think it's all mods that add custom generated structures and stuff in that nature, they are just exceeding a Tick Limit for generation that Mohist adds that we can't control at the moment. I know the properties config and the terraforged config have tick limit modifiers u can change, but some mods don't have that so they are at the mercy of whatever mohist's tick limit is, which currently from all the info i've gathered and read up on, is the issue. Hopefully a fix is somewhat soon, but since this project is open source, it's random for sure, but they seem to update it quite frequently!

**Comment by Won-Ton** [comment]
Snippet from the log on #816:
```
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1707)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3313)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1742)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture.join(CompletableFuture.java:1947)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.server.ServerChunkProvider.func_212849_a_(ServerChunkProvider.java:123)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.World.func_217353_a(World.java:246)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.IWorldReader.func_217348_a(IWorldReader.java:112)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.World.func_212866_a_(World.java:242)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.entity.Entity.func_242281_f(Entity.java:1309)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.entity.Entity.func_70080_a(Entity.java:1295)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: com.github.alexthe666.iceandfire.world.gen.WorldGenLightningDragonCave.func_241855_a(WorldGenLightningDragonCave.java:91)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: com.github.alexthe666.iceandfire.world.gen.WorldGenLightningDragonCave.func_241855_a(WorldGenLightningDragonCave.java:32)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.gen.feature.ConfiguredFeature.func_242765_a(SourceFile:55)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.biome.Biome.func_242427_a(Biome.java:254)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.gen.ChunkGenerator.func_230351_a_(ChunkGenerator.java:194)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.chunk.ChunkStatus.lambda$static$9(ChunkStatus.java:77)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.chunk.ChunkStatus$$Lambda$2351/391474186.doWork(Unknown Source)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.chunk.ChunkStatus.func_223198_a(ChunkStatus.java:198)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.server.ChunkManager.lambda$null$18(ChunkManager.java:552)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.server.ChunkManager$$Lambda$21727/1969626446.apply(Unknown Source)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: com.mojang.datafixers.util.Either$Left.map(Either.java:38)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.server.ChunkManager.lambda$chunkGenerate$20(ChunkManager.java:550)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.server.ChunkManager$$Lambda$21712/1277728948.apply(Unknown Source)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture.uniCompose(CompletableFuture.java:966)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:940)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:456)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.chunk.ChunkTaskPriorityQueueSorter.func_219083_b(SourceFile:58)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.world.chunk.ChunkTaskPriorityQueueSorter$$Lambda$21725/2049520254.run(Unknown Source)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.util.concurrent.DelegatedTaskExecutor.func_213148_e(DelegatedTaskExecutor.java:88)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.util.concurrent.DelegatedTaskExecutor.func_213145_a(DelegatedTaskExecutor.java:132)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: net.minecraft.util.concurrent.DelegatedTaskExecutor.run(DelegatedTaskExecutor.java:100)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1402)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)
[14:28:35] [Spigot Watchdog Thread/ERROR] [Minecraft/]: java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:175)
```

The deadlock is because IceAndFire has created an entity during generation of one of their structures, and that entity's init code is triggering a chunk request which can't be fullfilled because chunk generation is already blocked waiting for that structure to finish generating which in turn is waiting on that entity's request for a chunk (and so on).

Once the server has entered this state there is no recovery without killing it and restarting. Disabling the watchdog will only mean the server stays frozen indefinitely. I doubt any fault lies on this project, these deadlocks happen on plain forge servers too.

I don't know if it's the exact same problem in all the reports, but if you look out for similar-looking stacktraces you should be able to follow through and spot where exactly the deadlock is propagating from. As a rule of thumb, any use of a `World` object from within world-gen should set alarm bells ringing - each generation stage is given a 'region' to work inside to avoid precisely these kinds of issues (there'll be some exceptions I'm sure, but generally speaking, anyway).

**Comment by KowalskiXCIX** [comment]
Makes sense, i've been running a modpack with over 100+ mods and a lot of them add generated structures along with other mobs that go with those structures, (ice and fire is a perfect example) and i've never experienced a crash with 10 or so players running in the server generating structures, I decided to install mohist for testing because there are plugins that would make my server more "refined" in a sense, and I decided to create a new world, and that's when this issue began, i did tons and tons of testing the past week now and finally decided to post my findings here. This watchdog killing the server only occurs when mohist is installed, so it leads me to believe it's just a mod incompatibility thing, but it does it with a lot of other mods, so I wonder if it's something else in a sense. I don't code myself that much, I am not as knowledgable as you for sure, but atleast I have some evidence and some testing to help improve the methodology to find a fix :)

**Comment by VeritasSapere51** [comment]
I have encountered the same problem, the server fails to load in the chunks which presumably contain the dragon roosts after which the server hangs and crashes.

This is while only running iceandfire-2.1.6-1.16.4 and citadel-1.6.1.jar with Mohist 375

**Comment by KR33PYK1NG** [comment]
Fixed in a51e8c4
